Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp

```diff
@@ -160,9 +160,19 @@ objectmodel::BaseObject::SPtr ObjectFactory::createObject(objectmodel::BaseConte
     if (it != registry.end()) // Found the classname
     {
         entry = it->second;
-        // If no template has been given or if the template does not exist, first try with the default one
+        // If no template has been given or if the template does not exist, first try to infer one from the context
         if(templatename.empty() || entry->creatorMap.find(templatename) == entry->creatorMap.end())
-            templatename = entry->defaultTemplate;
+            templatename = entry->getPreferredTemplate(context, arg);
+
+        // If template was neither provided by user or deduced from context create the first one in the list.
+        if(templatename.empty() || entry->creatorMap.find(templatename) == entry->creatorMap.end())
+        {
+            for(auto& t : entry->creatorMap)
+            {
+                templatename = t.first;
+            }
+            templatename = entry->creatorMap.begin()->first;
```

Comment:
```suggestion
            templatename = entry->creatorMap.begin()->first;
```
The loop is not useful, right?